NetworkClientUtils.java

```diff
@@ -0,0 +1,196 @@
+
+    public NetworkClientPoller(long sleepTimeInMs) {
+      this.sleepTimeInMs = sleepTimeInMs;
+    }
+
+    public void run() {
+      while (!shutdown) {
+        try {
+          lock.lock();
+          List<RequestInfo> requestInfoListToSend = requestInfoList;
+          requestInfoList.clear();
+          lock.unlock();
+          List<ResponseInfo> responseInfos = networkClient.sendAndPoll(requestInfoListToSend, pollTimeMs);
+          onResponse(responseInfos);
+        } finally {
+          //lock.unlock();
+        }
+       /* try {
+          Thread.sleep(sleepTimeInMs);
+        } catch (InterruptedException e) {
+          e.printStackTrace();
+          shutdown = true;
+        }*/
+      }
+    }
+
+    void shutdown() {
+      this.shutdown = true;
+    }
+  }
+
+  /**
+   * Handle the response from polling the {@link NetworkClient}.
+   * @param responseInfoList the list of {@link ResponseInfo} containing the responses.
+   */
+  private void onResponse(List<ResponseInfo> responseInfoList) {
+    for (ResponseInfo responseInfo : responseInfoList) {
+      try {
+        RequestOrResponseType type = ((RequestOrResponse) responseInfo.getRequestInfo().getRequest()).getRequestType();
+        int correlationId = -1;
+        switch (type) {
+          case PutRequest:
+            correlationId = ((PutRequest) responseInfo.getRequestInfo().getRequest()).getCorrelationId();
+            break;
+          case GetRequest:
+            correlationId = ((GetRequest) responseInfo.getRequestInfo().getRequest()).getCorrelationId();
+            break;
+          case DeleteRequest:
+            correlationId = ((DeleteRequest) responseInfo.getRequestInfo().getRequest()).getCorrelationId();
+            break;
+          default:
+            logger.error("Unexpected response type: " + type + " received, discarding " + responseInfo);
+        }
+        if (correlationId != -1) {
+          RequestMetadata requestMetadata = correlationIdToRequestMetadataMap.remove(correlationId);
+          requestMetadata.futureResult.done(responseInfo.getResponse(),
+              responseInfo.getError() != null ? new NetworkClientException("The Operation could not be completed.",
+                  responseInfo.getError()) : null);
+          if (requestMetadata.callback != null) {
+            requestMetadata.callback.onCompletion(responseInfo.getResponse(),
+                responseInfo.getError() != null ? new NetworkClientException("The Operation could not be completed.",
+                    responseInfo.getError()) : null);
+          }
+        }
+      } catch (Exception e) {
+        logger.error("Unexpected error received while handling a response : " + responseInfo + ", exception : "
+            + e.getStackTrace());
+      }
+    }
+  }
+
+  /**
+   * Exception to hold the error thrown from the {@link NetworkClient}
+   */
+  static class NetworkClientException extends Exception {
+
+    private NetworkClientErrorCode errorCode;
```

Comment:
Yes, in next PR when we add metrics, will update respective error metrics. As of now, I don't do anything in specific wrt the errors thrown.